test/aws_test.go

```diff
@@ -4,32 +4,244 @@
+		},
+		{
+			helmKey: "pachd.storage.amazon.customEndpoint",
+			envVar:  "CUSTOM_ENDPOINT",
+			value:   "https://myfineendpoint.com",
+		},
+		{
+			helmKey: "pachd.storage.amazon.disableSSL",
+			envVar:  "DISABLE_SSL",
+			value:   "false",
+		},
+		{
+			helmKey: "pachd.storage.amazon.logOptions",
+			envVar:  "OBJ_LOG_OPTS",
+			value:   "", //TODO
+		},
+		/*{
+			helmKey: "pachd.storage.amazon.maxUploadParts",
+			envVar:  "MAX_UPLOAD_PARTS",
+			value:   "", //TODO
+		},*/
+		/*{
+			helmKey: "pachd.storage.amazon.verifySSL",
+			envVar:  "NO_VERIFY_SSL",
+			value:   "", //TODO
+		},*/
+		{
+			helmKey: "pachd.storage.amazon.partSize",
+			envVar:  "PART_SIZE",
+			value:   "", //TODO
+		},
+		/*{
+			helmKey: "pachd.storage.amazon.retries",
+			envVar:  "RETRIES",
+			value:   "", //TODO
+		},*/
+		/*{
+			helmKey: "pachd.storage.amazon.reverse",
+			envVar:  "REVERSE",
+			value:   "", //TODO
+		},*/
+		{
+			helmKey: "pachd.storage.amazon.timeout",
+			envVar:  "TIMEOUT",
+			value:   "", //TODO
+		},
+		{
+			helmKey: "pachd.storage.amazon.uploadACL",
+			envVar:  "UPLOAD_ACL",
+			value:   "", //TODO
+		},
+	}
+	var (
+		expectedServiceAccount = "my-fine-sa"
+		expectedProvisioner    = "ebs.csi.aws.com"
+		expectedStorageBackend = "AMAZON"
+	)
+
+	helmValues := map[string]string{
+		"pachd.storage.backend": "AMAZON",
+		`pachd.serviceAccount.additionalAnnotations.eks\.amazonaws\.com/role-arn`:  expectedServiceAccount,
+		`worker.serviceAccount.additionalAnnotations.eks\.amazonaws\.com/role-arn`: expectedServiceAccount,
+	}
+	for _, tc := range testCases {
+		helmValues[tc.helmKey] = tc.value
+	}
+
+	templatesToCheck := map[string]bool{
+		"templates/pachd/storage-secret.yaml":             false,
+		"templates/pachd/deployment.yaml":                 false,
+		"templates/pachd/rbac/serviceaccount.yaml":        false,
+		"templates/pachd/rbac/worker-serviceaccount.yaml": false,
+		"templates/etcd/statefulset.yaml":                 false,
+		"templates/etcd/storageclass-aws.yaml":            false,
+		"templates/postgresql/statefulset.yaml":           false,
+		"templates/postgresql/storageclass-aws.yaml":      false,
+	}
+
+	templatesToRender := []string{}
+	for k := range templatesToCheck {
+		templatesToRender = append(templatesToRender, k)
+	}
+
+	objects, err := manifestToObjects(helm.RenderTemplate(t, &helm.Options{
+		SetValues: helmValues,
+	}, "../pachyderm", "blah", templatesToRender))
+
 	if err != nil {
-		t.Error(err)
+		t.Fatal(err)
 	}
-	for _, obj := range objects {
-		switch obj := obj.(type) {
+
+	for _, object := range objects {
+		switch resource := object.(type) {
 		case *v1.Secret:
-			if obj.Name != "pachyderm-storage-secret" {
+			if resource.Name != "pachyderm-storage-secret" {
 				continue
 			}
-			if _, err := strconv.Atoi(string(obj.Data["part-size"])); err != nil {
-				t.Error(err)
+
+			for _, tc := range testCases {
+				t.Run(fmt.Sprintf("%s equals %s", tc.envVar, tc.value), func(t *testing.T) {
```

Comment:
I don't think it's worth t.Run for one test case.  It's good if you want to Fatal out of a logical test, or often run a subset of tests with "go test ... -run=TestFoo_Bar" (i.e. the other tests in TestFoo are slow), but it just adds extra boilerplate for these cases that are a single Errorf assertion.  I'd rewrite to:

```go
for _, rc := range testCases {
    if got, want := string(resource.Data[tc.envVar]), tc.value; got != want {
        t.Errorf("pachyderm-storage-secret: environment variable $%s:\n  got: %v\n want: %v", tc.envVar, got, want)
    }
}
```

If anything, the t.Run might go out a level of indentation and be called `pachyderm-storage-secret`:

```go
t.Run("pachyderm-storage-secret", func(t *testing.T) {
    for _, rc := range testCases {
        if got, want := string(resource.Data[tc.envVar]), tc.value; got != want {
            t.Errorf("environment variable $%s:\n  got: %v\n want: %v", tc.envVar, got, want)
        }
    }
}
```

For this case, both are good.  In general, the last example is best, I think.  Saves typing, and makes for very clear output.

This pattern comes up throughout the review; give each another look to make sure the extra level of indentation is worthwhile.

Reply:
Interesting, I thought sub tests were preferred as they add a distinct test to the output, making it more clear what your test cases are.